### ISSUE TITLE
research: can access to "set errored wf step to skip/complete" be restricted to repo admin?

### ISSUE THREAD

**Issue by ndushay** [sul-dlss/argo#1478]
Offshoot of #1346:  

> From an individual object WF view:
> * for errored steps: where we currently have "set to waiting", change it to a pulldown with “rerun, skip, complete” and hook these up to the correct stuff in the code
> * note that "rerun" is the same as what is now "set to waiting"

From [storytime](https://docs.google.com/document/d/1qmVpXy5ZA17EneeeRrIWUULpPTK9eh8s-53SVQ_Ockg) 2019-08-19, 

"Can we restrict skip/complete to the repo manager once we know which role that best maps to?"

"look into the possibilities with existing permissions and roles."

- do we have an existing role that is limited to the repository manager?
  - is `superadmin` such a role?
- can we easily restrict "skip" and "complete" status changes to only be allowed by a role?  For both:
  - individual workflow step of an item
  - workflow grid view
- can we easily fall back to the current display for non-repo manager?

**Comment by jmartin-sul** [comment]
possibly applicable workgroups/roles:
* `workgroup:sdr:manager-role` (the workgroup for the DOR manager role) (https://github.com/sul-dlss/argo/blob/master/app/models/user.rb#L10)
  * Argo `User#is_manager?` (https://github.com/sul-dlss/argo/blob/master/app/models/user.rb#L126-L129)
  * https://github.com/sul-dlss/argo/blob/master/app/models/ability.rb#L31
* `workgroup:sdr:administrator-role` (the "superadmin" DOR/argo role) (https://github.com/sul-dlss/argo/blob/master/app/models/user.rb#L9)
  * `User#is_webauth_admin?` (https://github.com/sul-dlss/argo/blob/master/app/models/user.rb#L118-L124) (i don't think we want to use `User#is_admin?`, because we want to restrict this to real admins, though in practice, if impersonation isn't buggy, only real super admins can use it anyway; still, better safe than sorry, imho)
  * https://github.com/sul-dlss/argo/blob/master/app/models/user.rb#L28

these are just cursory notes from my recollection of DOR/argo permissions, which i haven't worked with much recently.  the thing we need to do is to make sure that the usage of one of the above roles maps in practice to the set of users to whom we want to grant access to this new feature.  that is to say, if we use one of the above roles to gate access to this new feature, we should confirm that:
- [ ] the users in the underlying workgroup(s) for the role should all be granted access to the new feature
- [ ] anyone who isn't in said workgroup(s), but who we would add, should also have whatever permissions that workgroup/role grants elsewhere in argo.

if we can't confirm those assertions, we should create a new role, instead of re-using an existing one.  but my hope is that the manager role is appropriate, since the name seems to indicate that it's the right fit.

**Comment by ndushay** [comment]
thanks, @jmartin-sul - that's really helpful.  How do we find out which users are in a workgroup?

**Comment by jmartin-sul** [comment]
# some digging into DOR/argo manager and admin permissions


note: all search results below abbreviated to exclude things from specs, and to focus on what seemed like usage of the role in authorizing a user to do or view something.

## manager role

### workgroups for role
* `workgroup:sdr:manager-role`
  * users
    * _i don't have the requisite access to list these users (which is probably appropriate).  maybe ask andrew or ben?_

### where used in app

```
git grep is_manager # for User#is_manager?
```
* app/presenters/home_text_presenter.rb:  one of the roles that allows a user to view stuff
* app/search_builders/argo/access_controls_enforcement.rb:  one of the roles that allows user to view anything
* _**tangential question**:  should these permission checks go through cancan?_

```
git grep "can_manage?"  # question mark is just question mark unless it's -E grep; in practice, these are all Ability#can_manage?(ActiveFedora::Base))
```
* app/jobs/generic_job.rb:  provides a way to determine whether user can manage each druid in the batch
* app/jobs/manage_catkey_job.rb:  uses method from parent class on each druid to see whether user is allowed to update the catkey for it
* app/jobs/release_object_job.rb:  uses method from parent class on each druid to see whether user is allowed to release it
* app/jobs/set_governing_apo_job.rb:    `raise "user not authorized to move #{obj.pid} to #{new_apo_id}" unless ability.can?(:manage_governing_apo, obj, new_apo_id)`

```
git grep "authorize.*manage"
```
* apo_controller.rb, catalog_controller.rb, collections_controller.rb, collections_controller.rb, content_types_controller.rb, items_controller.rb
  * more individual object management

so, is allowed to "manage" any object in the system, which seems to mean view/version/edit.  but, unclear to me whether the user can get search results for all objects, or just objects that APOs grant them access to.  because i'm rusty on argo permission stuff, and am not sure i totally understand permitted_queries.rb.



## admin role

### workgroups for role
* `workgroup:sdr:administrator-role`
  * users
    * _sent to infra developers in slack_

### where used in app

```
git grep is_admin
```
* ability.rb:    `can :manage, :all if user.is_admin?`
  * so, any `can_manage?` or `authorize! :manage_FOO, @object` (where `FOO` is `item`, `content`, etc), or `authorize! :manage` call should return true if `User#is_admin?` is true for the user
* permitted_queries.rb:  if user is an admin, search results can include all objects, instead of filtering by objects governed by APOs that list the user
* home_text_presenter.rb:  one of the roles that allows a user to view stuff
* access_controls_enforcement.rb:  one of the roles that allows user to view anything

so definitely seems like this user can view/edit/etc everything, and can search for anything.




## webauth admin role

### workgroups for role
* workgroup:sdr:administrator-role
  * users: same as for the repository admin role.

### where used
```
git grep is_webauth_admin
```
* app/models/ability.rb: only thing that uses this directly is cancan definition for ability to impersonate
  * app/controllers/auth_controller.rb:    `authorize! :impersonate, User`

the difference in practice from the regular admin role is that `User#is_webauth_admin?` doesn't look at impersonated groups, whereas `User#is_admin?` does.  this allows the app to determine that a user really is still a super admin, for the purpose of presenting the button to clear impersonated groups, since `User#is_admin?` might no longer return `true` if the impersonated group list is the group list of a non-admin user.





# TLDR, recommendation

considering the above, i think i'd recommend using the `User#is_admin?` method to add an ability along the lines of:
```ruby
can :set_wf_step_to_any, :all if user.is_admin?
```

and then checking whether the user is authorized to set workflow steps to skipped/completed by doing something like:
```ruby
ability.can?(:set_wf_step_to_any)
```
or 
```ruby
authorize! :set_wf_step_to_any
```


it's been a while since i've worked with argo permissions or with cancan in general, so i'd appreciate it if someone would check my work.  IIRC, @jcoyne updated argo to use shib instead of webauth, so he may be the person with the most recent in-depth argo permissions experience?

**Comment by ndushay** [comment]
closing this research ticket;  have created #1516, #1517 to do this actual work.
